keps/sig-api-machinery/1623-standardize-conditions/README.md

```diff
@@ -0,0 +1,144 @@
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+While many Kuberentes APIs have `.status.conditions`, the schema of `condition` varies a lot between them.
+There is very little commonality at the level of serialization, proto-encoding, and required vs optional.
+Conditions are central enough to the API to make a common golang type with a fixed schema.
+The schema can be a strong recommendation to all API authors.
+
+## Motivation
+
+Allow general consumers to expect a common schema for `.status.conditions` and share golang logic for common Get, Set, Is for `.status.conditions`.
+The pattern is well-established and we have a good sense of the schema we now want.
+
+### Goals
+
+ 1. For all new APIs, have a common type for `.status.conditions`.
+ 2. Provide common utility methods for `HasCondition`, `IsConditionTrue`, `SetCondition`, etc.
+ 3. Provide recommended defaulting functions that set required fields and can be embedded into conversion/default functions.
+
+### Non-Goals
+
+ 1. Update all existing APIs to make use of the new condition type.
+
+## Proposal
+
+Introduce a type into k8s.io/apimachinery/pkg/apis/meta/v1 for `Condition` that looks like 
+```go
+type Condition struct {
+	// Type of condition in CamelCase.
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
+	// Last time the condition transitioned from one status to another.
+    // This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
```

Comment:
A side note (from a project that standardized this ourselves for our CRDs):

We found that using `metav1.Time` added a non-trivial amount of bookkeeping using a condition update semantic of "compute the new status, then update if any conditions changed value". To avoid this, we did the following:

1. Sort the Conditions by Type 
2. Compare the slices with `equality.Semantic`, ignoring LastTransitionTime
3. Only update Conditions which were not equal.

If you're curious, see https://github.com/knative/serving/pull/1663 and https://github.com/knative/pkg/blob/master/apis/volatile_time.go